ompi/datatype/ompi_datatype_module.c

```diff
@@ -545,8 +545,8 @@ int32_t ompi_datatype_init( void )
     do {                                                                \
         ompi_mpi_##name.dt.d_f_to_c_index =                             \
             opal_pointer_array_add(&ompi_datatype_f_to_c_table, &ompi_mpi_##name); \
-        if( ompi_datatype_number_of_predefined_data < (ompi_mpi_##name).dt.d_f_to_c_index ) \
```

Comment:
We cannot add the datatypes not in the monotonically increasing order with the currnet code because `opal_pointer_array_add` always returns a monotonically increasing value. So this check is redundant.

Reply:
We can be more specific here and instead of using opal_pointer_array_add we use the opal_pointer_array_set_item function when building the array of predefined types. We leave the registration of user datatypes (f2c) as is.

Reply:
@bosilca I changed the code to use `opal_pointer_array_set_item`. Is this what you mean? I didn't run enough test yet. If you are happy with this change, I'll run some tests and then merge this.